src/geometry/ArborX_GeometryTraits.hpp

```diff
@@ -0,0 +1,103 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+#ifndef ARBORX_GEOMETRY_TRAITS_HPP
+#define ARBORX_GEOMETRY_TRAITS_HPP
+
+#include <Kokkos_Core.hpp>
+
+namespace ArborX
+{
+
+namespace GeometryTraits
+{
+
+struct PointTag
+{};
+
+struct BoxTag
+{};
+
+struct SphereTag
+{};
+
+template <typename Geometry>
+struct dimension
+{
+  using not_specialized = void; // tag to detect existence of a specialization
+};
+
+template <typename Geometry>
+struct tag
+{
+  using not_specialized = void; // tag to detect existence of a specialization
+};
+
+template <typename Geometry>
+using DimensionNotSpecializedArchetypeAlias =
+    typename dimension<Geometry>::not_specialized;
+
+template <typename Geometry>
+using TagNotSpecializedArchetypeAlias = typename tag<Geometry>::not_specialized;
+
+template <typename Geometry>
+using DimensionArchetypeAlias = decltype(dimension<Geometry>::value);
+
+template <typename Geometry>
+using TagArchetypeAlias = typename tag<Geometry>::type;
+
+template <typename Geometry>
+struct is_point : std::is_same<Kokkos::detected_t<TagArchetypeAlias, Geometry>,
+                               PointTag>::type
+{};
+
+template <typename Geometry>
+struct is_box : std::is_same<Kokkos::detected_t<TagArchetypeAlias, Geometry>,
+                             BoxTag>::type
+{};
+
+template <typename Geometry>
+struct is_sphere : std::is_same<Kokkos::detected_t<TagArchetypeAlias, Geometry>,
+                                SphereTag>::type
+{};
+
+template <typename Geometry>
+void check_valid_geometry_traits()
+{
+  static_assert(
+      !Kokkos::is_detected<DimensionNotSpecializedArchetypeAlias, Geometry>{},
+      "Must specialize GeometryTraits::dimension<Geometry>");
+  static_assert(
+      !Kokkos::is_detected<TagNotSpecializedArchetypeAlias, Geometry>{},
+      "Must specialize GeometryTraits::tag<Geometry>");
+
+  static_assert(
+      Kokkos::is_detected<DimensionArchetypeAlias, Geometry>{},
+      "GeometryTraits::dimension<GeometrY> must define 'value' member type");
+  static_assert(
+      std::is_integral<
+          Kokkos::detected_t<DimensionArchetypeAlias, Geometry>>{} &&
+          GeometryTraits::dimension<Geometry>::value > 0,
```

Comment:
This will not work as you expect.
Consider
```
struct Foo {};
template <> struct dimension<Foo> { };
```
You will get a compile error because `GeometryTraits::dimension<Geometry>::value` gets "evaluated" even though `std::is_integral<...>::value` is `false`.

Reply:
Not sure understand. We have a check
```c++
  static_assert(
      Kokkos::is_detected<DimensionArchetypeAlias, Geometry>{},
      "GeometryTraits::dimension<Geometry> must define 'value' member type");
```
before this `static_assert`. So the `::value` is present. Uncommenting corresponding lines in `tstCompileOnlyGeometry.cpp` seems to do the right thing.

Reply:
You are right.  I overlooked it.